[NETBEANS-5627] [NETBEANS-5629] Gradle early Lookup and Classpath initialization.

sdedic opened the pull request:
This is an attempt to solve [NETBEANS-5627](https://issues.apache.org/jira/browse/NETBEANS-5627) and [NETBEANS-5629](https://issues.apache.org/jira/browse/NETBEANS-5629)

**It is still work in progress - but complete enough for feedback**

Major highlights of the change:
- a fresh new (untrusted) project still loads as `FALLBACK` state. **But** the Fallback heuristic is immediately applied, so the project (typically) knows its (java) source groups.
- creation attempts to load `EVALUATED` quality. This will eventually load cached info, if available
- with LegacyProjectLoader loading from above EVALUATED the project creation will not invoke Gradle build (the same as now).
- Lookup is initialized with plugins from the start: Fallback loader guesses something, and there's also [NETBEANS-5628](https://issues.apache.org/jira/browse/NETBEANS-5628) that could further improve the initial experience.
- Lookups are in a defined order: Plugins first, the default Lookup first, Plugins next, `_any` (any plugin, fallbacks) last. Similar to Maven. (see [NETBEANS-5629](https://issues.apache.org/jira/browse/NETBEANS-5629)). This allows to add pre-process, make specific stuff with plugin-specific services and do a post-process.
- If plugin Lookups change, their relative order is maintained
- ClassPath starts to listen at the start of its existence
- there were quite a few potential races in both NbGradleProjectImpl and ClasspathProviderImpl. Attempted to synchronize, while firing events out outside of any locks.

In effect, a code that does
```
FileObject someSource = ...;
ClassPath cp = ClassPath.getClassPath(someSource);
```
starts working even on an **unopened** Gradle project, typically. In less typical setups one has to prime / build the project, so NetBeans create its caches.

There are things to improve still:
- apichanges: the Lookups need to be documented, but see the discussion in NETBEANS-5629. Also GradleProject has one more method to (re)load the project at the defined quality
- classpath queries **might** attempt to upgrade the project to FULL when it realizes the project is lesser quality. But I didn't want to do it synchronously in the query - maybe should be done asynchronosly, but that's still lacking some reload pooling
- an attempted reload could piggyback on an asynchronous **pending or scheduled** reload. Typically a reload could reuse stuff from reload-after-build or a pending upgrade-quality reload.
- I must write more tests.

lkishalmi commented on extide/gradle/src/org/netbeans/modules/gradle/NbGradleProjectImpl.java:
```diff
@@ -223,52 +229,179 @@ void detachAllUpdater() {
     }
 
-    public Quality getAimedQuality() {
+    public final Quality getAimedQuality() {
```
NbGradleProjectImpl class is already final.

lkishalmi commented on java/gradle.java/src/org/netbeans/modules/gradle/java/classpath/ClassPathProviderImpl.java:
```diff
@@ -70,36 +71,58 @@
+    
+    private void updateResources(URI uri) {
+        if ((uri != null) && (uri.getPath() != null) && uri.getPath().endsWith(MODULE_INFO_JAVA)) {
+            GradleJavaProject gjp = GradleJavaProject.get(ClassPathProviderImpl.this.project);
```
ClassPathProviderImpl.this.project -> project ?

lkishalmi reviewed (approved):
Really nice work! Thank you!
The changes seem to be reasonable, can't wait to test this out in my local build.
Mentioned only two stylistic thing, but those do not really matter.

victorwss commented:
Sorry to comment this here, but since you (@sdedic) are working in the issue and the mailing lists are a confusing mess that I'm unable to follow, let me comment here. Pardon me if you already know what I will tell.

I started to try out Netbeans 12.3 + Gradle 7 + Java 16 and the result was very bad. The machine was fresh and there was no Java nor Netbeans ever installed before. So I installed Java 16 and then just downloaded and installed Netbeans 12.3.

I got a small project in Gradle and tried to build it. My first result was a silent NullPointerExceptions somewhere inside Netbeans when it tried to access some array in order to build the project, with the output panel blank. If you are interested, I can backtrack to this point in order to reproduce it.

Anyway, trying to make it work, I manually downloaded Gradle 7.0 and setted it in the Options > Java > Gradle setting.

Then, when Netbeans tries to to the prime-loading the project with its internally downloaded Gradle 6.7 regardless of any setting that I to try to do to force the use of Gradle 7.0. This results in a Gradle instantly crashing with unsupported classfile format because Gradle 6.7 can't recognize Java 16.

To make it worse, when Netbeans fails in the prime-loading, it alerts that the project is unloadable due to errors and fallbacks to an unchangeable "-source 5" which makes the java files are full of spurious errors about uses of diamonds, lambdas, records and a lot of other stuffs.

The result is that so far, is that Netbeans 12.3 + Gradle 7 + Java 16 is pretty much an unusable combination. I lost some hours trying to make this work and after searching a lot on Google, I just landed here.

So, I just wanted to inform you about the experience in order to avoid that some of those cases are missed. And also, to tell you that this problem is surely seen in the wild and are biting some random people out there like me. Anyway, keep the good work!

sdedic commented:
@victorwss Thanks for your feedback. Sadly this particular PR is probably completely unrelated to gradle version + Java issues. 
This PR aims to improve NetBeans' tools cooperation with the gradle module; it never aimed to fix anything between NB gradle module and the gradle daemon itself.

If you can reproduce the defect(s), then great ! Please file JIRA issues for each one separately; if you manage to get a stacktrace of the error, please attach it to that issue. Include reproduction steps; ideally starting from a fresh install or project open. GitHub PRs discussions are not intended for reporting (general) bugs, or for product feedback, the primary purpose is to review and discuss changes proposed in the pull request.

lkishalmi commented:
@victorwss Use NetBeans 12.4 (rc1 is out rc2 is coming) that has Gadle 7.0 support.

sdedic commented:
@victorwss Incidentally, I had some internal (*wink*) report of the **same** issue that I needed to process. See [micronaut-starter#757](https://github.com/micronaut-projects/micronaut-starter/issues/757)

sdedic commented:
Added some tests, changed `toQuality` to return just `CompletionStage` plus improved `ClassPathProviderImpl` to make the project to initialize (if it was not yet).

@lkishalmi  please re-review

JaroslavTulach commented on extide/gradle/src/org/netbeans/modules/gradle/api/NbGradleProject.java:
```diff
@@ -206,6 +210,30 @@ public Quality getAimedQuality() {
+     * to block waiting for the result.
+     * @since 2.11
+     */
+    public @NonNull CompletionStage<NbGradleProject> toQuality(@NullAllowed String reason, @NonNull Quality q, boolean forceLoad) {
```
Great to see `CompletionStage` in out APIs! I'd say `CompletionStage` is the best after invention of sliced bread!

lkishalmi replied:
Interesting way to move forward! Really waiting for this PR to go in mainline after the 12.4 API snapshot happens.
